pkg/sidekick/adhocprofiles/adhocprofiles.go

```diff
@@ -0,0 +1,247 @@
+package adhocprofiles
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	"connectrpc.com/connect"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/tenant"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+
+	v1 "github.com/grafana/pyroscope/api/gen/proto/go/adhocprofiles/v1"
+	"github.com/grafana/pyroscope/pkg/frontend"
+	"github.com/grafana/pyroscope/pkg/objstore"
+	"github.com/grafana/pyroscope/pkg/og/convert"
+	"github.com/grafana/pyroscope/pkg/og/structs/flamebearer"
+	"github.com/grafana/pyroscope/pkg/validation"
+)
+
+type AdHocProfiles struct {
+	logger    log.Logger
+	limits    frontend.Limits
+	bucket    objstore.Bucket
+	buckets   map[string]objstore.Bucket
+	bucketsMu sync.RWMutex
+}
+
+type AdHocProfile struct {
+	Name       string    `json:"name"`
+	Data       string    `json:"data"`
+	UploadedAt time.Time `json:"uploadedAt"`
+}
+
+func NewAdHocProfiles(bucket objstore.Bucket, logger log.Logger, limits frontend.Limits) *AdHocProfiles {
+	a := &AdHocProfiles{
+		logger:  logger,
+		bucket:  bucket,
+		buckets: make(map[string]objstore.Bucket),
+		limits:  limits,
+	}
+	return a
+}
+
+func (a *AdHocProfiles) Upload(ctx context.Context, c *connect.Request[v1.AdHocProfilesUploadRequest]) (*connect.Response[v1.AdHocProfilesGetResponse], error) {
+	tenantID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+
+	adHocProfile := AdHocProfile{
+		Name:       c.Msg.Name,
+		Data:       c.Msg.Profile,
+		UploadedAt: time.Now().UTC(),
+	}
+
+	// TODO: Add per-tenant upload limits (number of files, total size, etc.)
+
+	maxNodes, err := validation.ValidateMaxNodes(a.limits, []string{tenantID}, c.Msg.GetMaxNodes())
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not determine max nodes")
+	}
+
+	profile, profileTypes, err := parse(&adHocProfile, nil, maxNodes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse profile")
+	}
+
+	bucket, err := a.getBucket(tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	uid := ulid.MustNew(ulid.Timestamp(adHocProfile.UploadedAt), rand.Reader)
+	id := strings.Join([]string{uid.String(), adHocProfile.Name}, "-")
```

Comment:
Keys in the object storage will have the form `01HMXV8BF4EH71NBYZNPPVGJ2X-cpu.pprof`, using `ULID` for the first segment to simplify embedding a timestamp. This is then used in the `/List` endpoint to return meaningful data without querying each individual profile.